esutil/bulk_indexer.go

```diff
@@ -0,0 +1,509 @@
+// Licensed to Elasticsearch B.V. under one or more agreements.
+// Elasticsearch B.V. licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information.
+
+package esutil
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"runtime"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/elastic/go-elasticsearch/v8"
+	"github.com/elastic/go-elasticsearch/v8/esapi"
+)
+
+// BulkIndexer represents a parallel, asynchronous, efficient indexer for Elasticsearch.
+//
+type BulkIndexer interface {
+	// Add adds an item to the indexer. It returns an error when the item cannot be added.
+	// Use the OnSuccess and OnFailure callbacks to get the operation result for the item.
+	//
+	// You must call the Close() method after you're done adding items.
+	//
+	// It is safe for concurrent use. When it's called from goroutines,
+	// they must finish before the call to Close, eg. using sync.WaitGroup.
+	Add(context.Context, BulkIndexerItem) error
+
+	// Close waits until all added items are flushed and closes the indexer.
+	Close(context.Context) error
+
+	// Stats returns indexer statistics.
+	Stats() BulkIndexerStats
+}
+
+// BulkIndexerConfig represents configuration of the indexer.
+//
+type BulkIndexerConfig struct {
+	NumWorkers    int           // The number of workers. Defaults to runtime.NumCPU().
+	FlushBytes    int           // The flush threshold in bytes. Defaults to 5MB.
+	FlushInterval time.Duration // The flush threshold as duration. Defaults to 10sec.
+
+	Client      *elasticsearch.Client   // The Elasticsearch client.
+	Decoder     BulkResponseJSONDecoder // A custom JSON decoder.
+	DebugLogger BulkIndexerDebugLogger  // An optional logger for debugging.
+
+	// Parameters of the Bulk API.
+	Index               string
+	ErrorTrace          bool
+	FilterPath          []string
+	Header              http.Header
+	Human               bool
+	Pipeline            string
+	Pretty              bool
+	Refresh             string
+	Routing             string
+	Source              []string
+	SourceExcludes      []string
+	SourceIncludes      []string
+	Timeout             time.Duration
+	WaitForActiveShards string
+}
+
+// BulkIndexerStats represents the indexer statistics.
+//
+type BulkIndexerStats struct {
+	NumAdded    uint
+	NumFlushed  uint
+	NumFailed   uint
+	NumIndexed  uint
+	NumCreated  uint
+	NumUpdated  uint
+	NumDeleted  uint
+	NumRequests uint
+}
+
+// BulkIndexerItem represents an indexer item.
+//
+type BulkIndexerItem struct {
+	Index           string
+	Action          string
+	DocumentID      string
+	Body            io.Reader
+	RetryOnConflict *int
+
+	Metadata interface{} // To use eg. in OnSuccess/OnFailure callbacks
+
+	OnSuccess func(item BulkIndexerItem, res BulkIndexerResponseItem)            // Per item
+	OnFailure func(item BulkIndexerItem, res BulkIndexerResponseItem, err error) // Per item
```

Comment:
Isn't that what I suggested? i.e. the BulkIndexer accepts a channel through the BulkIndexerItem. However, it's not unbounded: it's 0 or 1.

Reply:
Ah right, sorry, I misread the code.

I was thinking about it a little bit more, and I'm leaning towards keeping the notifications as callbacks. In a talk linked from the article cited above, Alan Schreve makes a good point — at around [16:50](https://youtu.be/hFqXgmor74k?t=1010) —, that a callback-type of notification can be easily turned into a channel-based notification, but not vice versa. I'm thinking about something like:

```golang
func(item BulkIndexerItem, res BulkIndexerResponseItem) {
	updatedIDs <- item.DocumentID
}
```

I've played with it a little bit in the unit tests, pushed to a separate branch here: https://github.com/elastic/go-elasticsearch/commit/706ad08e730101d898fdc39beb5f7aaf4abae352.

Reply:
I'd prefer to have a callback. As @karmi said, I can easily add channels with semantics I need. E.g. 
- a channel could be global on the Indexer
- two per global Indexer (success, failure) reporting the IDs that have been indexed or failed
- one per Item. If per item do I have to create it and the API will close it for me? If so I can easily do this via callbacks without enforcing a 'signaling' protocol between user and API.

Reply:
Have you considered one single callback like `OnProcessed(BulkIndexerItem, BulkIndexerResponseItem)` (with the `BulkIndexerResponseItem` also including an error value)? This reduces the number of callbacks one has to implement (at the potential cost of an if-else).

Reply:
Thanks for the comment, @urso! The more I think about it, the more I think the callback-based approach is both safer for the calling code, as well as more flexible (as outlined above). But I'm open to further discussion, of course.

Reply:
> Have you considered one single callback like `OnProcessed` (...)

That's an interesting idea, to combine everything into a single "handler". I can see how this would simplify the calling code, but, as you note, it would have to need to differentiate between success and failure inside the callback.

This could get rather complex, like checking for `err != nil`, then checking if `res.IsError()`, and so on... But there's something intriguing about an API like that — I'll think about it some more.

Reply:
I know we can wrap a channel with functions. I'm asking the question to get a feel for what we're optimising for, e.g. do the callers of "Add" typically want to wait for the bulk response? If so, a channel will be most useful. For other cases where you want an observer style interface where you're notified of all responses, then I think a callback would be fine.

That is to say, I think there's room for two mechanisms:
 - a callback/interface configured on the BulkIndexer itself, which is invoked for each request/response
 - a callback/interface/channel on the item, which is invoked on that item's response

The first one would be useful for things like metrics, which won't significantly vary the callback behaviour based on the item.

It might also be worth considering an interface, rather than choosing either a function or channel. Implementations of both could be provided. e.g.

```go
type BulkResponse struct {
    Item BulkIndexerItem
    Err error // If Err is non-nil, Response is undefined
    Response BulkIndexerResponseItem
}

type BulkResponseHandler interface {
    HandleBulkResponse(BulkResponse)
}

type BulkResponseHandlerFunc func(BulkResponse)
func (f BulkResponseHandlerFunc) HandleBulkResponse(...) {...}

type BulkResponseHandlerChan chan<- BulkResponse
func (ch BulkResponseHandlerChan) HandleBulkResponse(...) {...}
```

Just a thought.

+1 to combining OnSuccess/OnFailure anyway.

Reply:
> I'm asking the question to get a feel for what we're optimising for, e.g. do the callers of "Add" typically want to wait for the bulk response?

I would expect most callers are interested in an answer to the question "what has happened to my item", most commonly creating a document. In case that operation wasn't successful, they would want to put that item into some sort of dead-letter queue, for inspection and re-processing.

The details of the HTTP response itself (ie. the result of a request to `_bulk`) is not something terribly interesting to the caller, I imagine. That's why I have designed the callbacks to be per-item since the start, differently from the Java `BulkProcessor` class or its Go port in `olivere/elastic`. (I had callbacks for these events as well, `OnBatchSuccess()` etc, but removed them during development.)

> It might also be worth considering an interface (...)

In general, I'm a big proponent of putting an implementation behind an interface. In this case, however, I feel like it introduces too much ceremony around a simple `func`-based callback. With the `func`-based callback, the functionality is clear from reading a godoc output or scanning the source code. Also, those additional types add more weight to an already big public API, and since this component lives in the rather generic `esutil` package, it might become quite big. (I've considered putting the component into a subpackage, something like `esbulk`, but wasn't sure if it wouldn't impact the discoverability of the component too much. The new [godoc website](https://pkg.go.dev/github.com/elastic/go-elasticsearch/v7?tab=doc) makes it arguably even harder to discover "Subdirectories"...)

> +1 to combining OnSuccess/OnFailure anyway.

Can you clarify why would you do this? As mentioned above, this shifts the responsibility of detecting the result of the operation to the caller. Sure, it's "just an `if/else`", mostly. But to me, it's more about _defining the semantics_ of what constitutes a success and error. Consider this code (which Silvia has asked about):

https://github.com/elastic/go-elasticsearch/blob/6265c4b76aee00190e6d7de45495e69cb39cafaa/esutil/bulk_indexer.go#L477

This is handling eg. this result in the Bulk API response, which indicates that a document to be deleted wasn't found:

https://github.com/elastic/go-elasticsearch/blob/6265c4b76aee00190e6d7de45495e69cb39cafaa/esutil/testdata/bulk_response_1b.json#L22-L34

Notice that the response JSON has `"errors": false` in the top-level field, and that there's no `error` for the individual item. However, deleting a document which wasn't found constitutes an error to me, and I would expect the `OnFailure()` callback to fire. This type of knowledge requires deep understanding how Elasticsearch behaves, and it took myself some time to understand what's going on, and a lot of confusion about why this operation isn't marked with an `error`. Therefore, I would like to depend on a high-level component such as this one to make these informed decisions.

Reply:
> Can you clarify why would you do this? As mentioned above, this shifts the responsibility of detecting the result of the operation to the caller. Sure, it's "just an if/else", mostly

If having to write callbacks, personally I'd prefer to write only one callback with if-else instead of two separate callbacks.

Using interfaces is a nice idea, in case we want to have multiple callbacks.

For both, the callback approach (having a single callback) and the interfaces approach we can introduce constructors wrapping different use-cases. E.g.

```
ch := make(chan struct{})
...

BulkIndexerResponseItem{
  ...
  OnProcessed: esutil.Notify(ch)
}
```

with:

```
func Notify(ch chan struct{}) BulkResponder {
  return func(...) {
    close(ch)
  }
}
```

I'd say having a function is not that much different from an interface, as in either case we might want to provide additional helpers.

Reply:
Thanks for all the thoughts. At the moment, I'm more or less on the side of the fence where:

* `func`-style callbacks are used for notification,
* there are two separate callbacks, for success and failure.

Regarding the separate callbacks, I can't force myself to merge them into one... Among conceptual problems, I struggle with the naming. `OnProcessed()` seems wrong to me — if there's an error, the item hasn't really been "processed" at all.

Also, I've added a `BulkIndexerConfig.OnError()` callback in 9ec0d6b, to properly propagate errors from the depth of the indexer, and notify the calling code. Can you please look at that?

Reply:
In the end, we went full length with callbacks, adding also `OnFlushStart()` and `OnFlushEnd()`. Many thanks for the conversation again!